README.md
# SK-Q-Navigator__VB6
Bahan Ajar Fundamental Pemrograman Visual Basic 6.0 - Studi Kasus; Q Navigator by U.Sudeep Nayak, Randy Birch, NeoText, Night Wolf<br><br>
<img src="https://github.com/RizkyKhapidsyah/SK-Q-Navigator__VB6/blob/main/Result/001.PNG"><br><br>
Lihat Source Code : <br>
- <a href="https://github.com/RizkyKhapidsyah/SK-Q-Navigator__VB6">Program</a>


history.txt
Monday
	about:blank
		about:blank/
	www.sjbhs2000.com
		http://www.sjbhs2000.com/
	www.planet-source-code.com
		http://www.planet-source-code.com/xq/ASP/txtCodeId.13997/lngWId.1/qx/vb/scripts/ShowCode.htm
Wednesday
	about:blank
		about:blank/
Friday


Module1.bas
Attribute VB_Name = "Module1"
Option Explicit

Public Const sFileCount As String = "###,###,###,##0 f\i\l\e\s\ \f\o\u\n\d"
Public Const ERROR_CACHE_FIND_FAIL As Long = 0
Public Const ERROR_CACHE_FIND_SUCCESS As Long = 1
Public Const ERROR_FILE_NOT_FOUND As Long = 2
Public Const ERROR_ACCESS_DENIED As Long = 5
Public Const ERROR_INSUFFICIENT_BUFFER As Long = 122
Public Const MAX_PATH  As Long = 260
Public Const MAX_CACHE_ENTRY_INFO_SIZE As Long = 4096

Public Const LMEM_FIXED As Long = &H0
Public Const LMEM_ZEROINIT As Long = &H40
Public Const LPTR As Long = (LMEM_FIXED Or LMEM_ZEROINIT)

Public Const NORMAL_CACHE_ENTRY As Long = &H1
Public Const EDITED_CACHE_ENTRY As Long = &H8
Public Const TRACK_OFFLINE_CACHE_ENTRY As Long = &H10
Public Const TRACK_ONLINE_CACHE_ENTRY As Long = &H20
Public Const STICKY_CACHE_ENTRY As Long = &H40
Public Const SPARSE_CACHE_ENTRY As Long = &H10000
Public Const COOKIE_CACHE_ENTRY As Long = &H100000
Public Const URLHISTORY_CACHE_ENTRY As Long = &H200000
Public Const URLCACHE_FIND_DEFAULT_FILTER As Long = NORMAL_CACHE_ENTRY Or _
                                                   COOKIE_CACHE_ENTRY Or _
                                                   URLHISTORY_CACHE_ENTRY Or _
                                                   TRACK_OFFLINE_CACHE_ENTRY Or _
                                                   TRACK_ONLINE_CACHE_ENTRY Or _
                                                   STICKY_CACHE_ENTRY
Public Type FILETIME
   dwLowDateTime As Long
   dwHighDateTime As Long
End Type

Public Type INTERNET_CACHE_ENTRY_INFO
   dwStructSize As Long
   lpszSourceUrlName As Long
   lpszLocalFileName As Long
   CacheEntryType  As Long
   dwUseCount As Long
   dwHitRate As Long
   dwSizeLow As Long
   dwSizeHigh As Long
   LastModifiedTime As FILETIME
   ExpireTime As FILETIME
   LastAccessTime As FILETIME
   LastSyncTime As FILETIME
   lpHeaderInfo As Long
   dwHeaderInfoSize As Long
   lpszFileExtension As Long
   dwExemptDelta  As Long
End Type

Public Declare Function FindFirstUrlCacheEntry Lib "Wininet.dll" _
   Alias "FindFirstUrlCacheEntryA" _
  (ByVal lpszUrlSearchPattern As String, _
   lpFirstCacheEntryInfo As Any, _
   lpdwFirstCacheEntryInfoBufferSize As Long) As Long

Public Declare Function FindNextUrlCacheEntry Lib "Wininet.dll" _
   Alias "FindNextUrlCacheEntryA" _
  (ByVal hEnumHandle As Long, _
   lpNextCacheEntryInfo As Any, _
   lpdwNextCacheEntryInfoBufferSize As Long) As Long

Public Declare Function FindCloseUrlCache Lib "Wininet.dll" _
   (ByVal hEnumHandle As Long) As Long
   
Public Declare Sub CopyMemory Lib "kernel32" _
   Alias "RtlMoveMemory" _
   (pDest As Any, _
    pSource As Any, _
    ByVal dwLength As Long)

Public Declare Function lstrcpyA Lib "kernel32" _
  (ByVal RetVal As String, ByVal Ptr As Long) As Long
                        
Public Declare Function lstrlenA Lib "kernel32" _
  (ByVal Ptr As Any) As Long
  
Public Declare Function LocalAlloc Lib "kernel32" _
   (ByVal uFlags As Long, _
    ByVal uBytes As Long) As Long
    
Public Declare Function LocalFree Lib "kernel32" _
   (ByVal hMem As Long) As Long
'--end block--'
   






Module3.bas
Attribute VB_Name = "Module3"
Global lLeft, lTop, lWidth, lHeight As Long
Public Function sudpre()
If Form1.TabStrip1.SelectedItem.index = 1 Then
Form1.TabStrip1.Tabs(Form1.TabStrip1.Tabs.Count).Selected = True
Else
Form1.TabStrip1.Tabs(Form1.TabStrip1.SelectedItem.index - 1).Selected = True
    Form1.TabStrip1.SetFocus
    Form1.TabStrip1.Refresh
End If
End Function
Public Function sudnex()
If Form1.TabStrip1.SelectedItem.index = Form1.TabStrip1.Tabs.Count Then
Form1.TabStrip1.Tabs(1).Selected = True
Else
Form1.TabStrip1.Tabs(Form1.TabStrip1.SelectedItem.index + 1).Selected = True
    Form1.TabStrip1.SetFocus
    Form1.TabStrip1.Refresh
End If
End Function


Sub RepositionProgressBar()
        'RESIZE and POSITION THE PROGRESS BAR
        lLeft = Form1.StatusBar.Panels(1).Left + 10
        lTop = Form1.StatusBar.Top + 80
        lWidth = Form1.StatusBar.Panels(1).Width - 20
        lHeight = Form1.StatusBar.Height - 70
        Form1.ProgressBar1.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar2.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar3.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar4.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar5.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar6.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar7.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar8.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar9.Move lLeft, lTop, lWidth, lHeight
Form1.ProgressBar10.Move lLeft, lTop, lWidth, lHeight

End Sub



Module5.bas
Attribute VB_Name = "Module5"
'How to detect if another application has established
'the connection to Internet.
'Run Only in WIN95,98 (not tested in NT)
'If the connection is succesfull, the function
'IsConnected return True
'The Burn Poodle

Option Explicit
Public Declare Function RasEnumConnections Lib "RasApi32.dll" Alias "RasEnumConnectionsA" (lpRasCon As Any, lpcb As Long, lpcConnections As Long) As Long
Public Declare Function RasGetConnectStatus Lib "RasApi32.dll" Alias "RasGetConnectStatusA" (ByVal hRasCon As Long, lpStatus As Any) As Long
'
Public Const RAS95_MaxEntryName = 256
Public Const RAS95_MaxDeviceType = 16
Public Const RAS95_MaxDeviceName = 32
'
Public Type RASCONN95
    dwSize As Long
    hRasCon As Long
    szEntryName(RAS95_MaxEntryName) As Byte
    szDeviceType(RAS95_MaxDeviceType) As Byte
    szDeviceName(RAS95_MaxDeviceName) As Byte
End Type
'
Public Type RASCONNSTATUS95
    dwSize As Long
    RasConnState As Long
    dwError As Long
    szDeviceType(RAS95_MaxDeviceType) As Byte
    szDeviceName(RAS95_MaxDeviceName) As Byte
End Type
        


Public Function IsConnected() As Boolean
On Error Resume Next
Dim TRasCon(255) As RASCONN95
Dim lg As Long
Dim lpcon As Long
Dim RetVal As Long
Dim Tstatus As RASCONNSTATUS95
'
TRasCon(0).dwSize = 412
lg = 256 * TRasCon(0).dwSize
'
RetVal = RasEnumConnections(TRasCon(0), lg, lpcon)
If RetVal <> 0 Then
                    MsgBox "ERROR"
                    Exit Function
                    End If
'
Tstatus.dwSize = 160
RetVal = RasGetConnectStatus(TRasCon(0).hRasCon, Tstatus)
If Tstatus.RasConnState = &H2000 Then
                         IsConnected = True
                         Else
                         IsConnected = False
                         End If

End Function



Module7.bas
Attribute VB_Name = "Module7"
Option Explicit

Public Declare Function DoAddToFavDlg Lib "shdocvw.dll" _
  (ByVal hwnd As Long, _
   ByVal szPath As String, _
   ByVal nSizeOfPath As Long, _
   ByVal szTitle As String, _
   ByVal nSizeOfTitle As Long, _
   ByVal pidl As Long) As Long
   
Public Declare Function DoOrganizeFavDlg Lib "shdocvw.dll" _
  (ByVal hwnd As Long, _
   ByVal lpszRootFolder As String) As Long


Public Const MAXDWORD = &HFFFFFFF
Public Const MAX_PATH = 260
Public Const INVALID_HANDLE_VALUE = -1
Public Const FILE_ATTRIBUTE_ARCHIVE = &H20
Public Const FILE_ATTRIBUTE_COMPRESSED = &H800
Public Const FILE_ATTRIBUTE_DIRECTORY = &H10
Public Const FILE_ATTRIBUTE_HIDDEN = &H2
Public Const FILE_ATTRIBUTE_NORMAL = &H80
Public Const FILE_ATTRIBUTE_READONLY = &H1
Public Const FILE_ATTRIBUTE_TEMPORARY = &H100
Public Const FILE_ATTRIBUTE_FLAGS = FILE_ATTRIBUTE_ARCHIVE Or _
                                     FILE_ATTRIBUTE_HIDDEN Or _
                                     FILE_ATTRIBUTE_NORMAL Or _
                                     FILE_ATTRIBUTE_READONLY

Public Const SHGFP_TYPE_CURRENT As Long = &H0
Public Const SHGFP_TYPE_DEFAULT As Long = &H1
Public Const CSIDL_FAVORITES As Long = &H6
Public Const CSIDL_COMMON_FAVORITES As Long = &H1F
Public Const MAX_LENGTH As Long = 260
Public Const S_OK As Long = 0
Public Const S_FALSE As Long = 1


Public Const DRIVE_UNKNOWNTYPE = 1
Public Const DRIVE_REMOVABLE = 2
Public Const DRIVE_FIXED = 3
Public Const DRIVE_REMOTE = 4
Public Const DRIVE_CDROM = 5
Public Const DRIVE_RAMDISK = 6

Public Type FILETIME
   dwLowDateTime As Long
   dwHighDateTime As Long
End Type

Public Type WIN32_FIND_DATA
   dwFileAttributes As Long
   ftCreationTime As FILETIME
   ftLastAccessTime As FILETIME
   ftLastWriteTime As FILETIME
   nFileSizeHigh As Long
   nFileSizeLow As Long
   dwReserved0 As Long
   dwReserved1 As Long
   cFileName As String * MAX_PATH
   cAlternate As String * 14
End Type

'the custom-made UDT for searching
Public Type FILE_PARAMS
   bRecurse As Boolean     'set True to perform a recursive search
   bList As Boolean        'set True to add results to listbox
   bFound As Boolean       'set only with SearchTreeForFile methods
   sFileRoot As String     'search starting point, ie c:\, c:\winnt\
   sFileNameExt As String  'filename/filespec to locate, ie *.dll, Q - Pad.exe
   sResult As String       'path to file. Set only with SearchTreeForFile methods
   nFileCount As Long      'total file count matching filespec. Set in FindXXX only
   nFileSize As Double     'total file size matching filespec. Set in FindXXX only
End Type

Public Declare Function FindClose Lib "kernel32" _
  (ByVal hFindFile As Long) As Long
   
Public Declare Function FindFirstFile Lib "kernel32" _
   Alias "FindFirstFileA" _
  (ByVal lpFileName As String, _
   lpFindFileData As WIN32_FIND_DATA) As Long
   
Public Declare Function FindNextFile Lib "kernel32" _
   Alias "FindNextFileA" _
  (ByVal hFindFile As Long, _
   lpFindFileData As WIN32_FIND_DATA) As Long

Public Declare Function GetLogicalDriveStrings Lib "kernel32" _
   Alias "GetLogicalDriveStringsA" _
  (ByVal nBufferLength As Long, ByVal lpBuffer As String) As Long
      
Public Declare Function GetDriveType Lib "kernel32" _
   Alias "GetDriveTypeA" _
  (ByVal nDrive As String) As Long

Public Declare Function GetDesktopWindow Lib "user32" () As Long

Public Const SW_SHOWNA = 8
Public Const SW_SHOWNORMAL = 1
Public Const SW_SHOWDEFAULT = 10


Public Declare Function ShellExecute Lib "shell32.dll" _
   Alias "ShellExecuteA" _
  (ByVal hwnd As Long, ByVal lpOperation As String, _
   ByVal lpFile As String, ByVal lpParameters As String, _
   ByVal lpDirectory As String, ByVal nShowCmd As Long) As Long

Public Declare Function FindExecutable Lib "shell32.dll" Alias _
         "FindExecutableA" (ByVal lpFile As String, ByVal lpDirectory As _
         String, ByVal lpResult As String) As Long
   

Public Declare Function GetPrivateProfileString _
   Lib "kernel32" Alias "GetPrivateProfileStringA" _
  (ByVal lpSectionName As String, _
   ByVal lpKeyName As Any, _
   ByVal lpDefault As String, _
   ByVal lpReturnedString As String, _
   ByVal nSize As Long, _
   ByVal lpFileName As String) As Long
   
Public Declare Function SHGetFolderPath _
    Lib "shfolder.dll" Alias "SHGetFolderPathA" _
   (ByVal hwndOwner As Long, _
     ByVal nFolder As Long, _
     ByVal hToken As Long, _
     ByVal dwReserved As Long, _
     ByVal lpszPath As String) As Long
      

Type FavMenu
    FavName As String
    FavURL As String
    FavRoot As String
    FavFolder As String
End Type


Private Sub GetFileInformation(FP As FILE_PARAMS)

  'local working variables
   Dim WFD As WIN32_FIND_DATA
   Dim hFile As Long
   Dim sPath As String
   Dim sRoot As String
   Dim sTmp As String

  'FP.sFileRoot (assigned to sRoot) contains
  'the path to search.
  
  'FP.sFileNameExt (assigned to sPath) contains
  'the full path and filespec.
   sRoot = QualifyPath(FP.sFileRoot)
   sPath = sRoot & FP.sFileNameExt

  'obtain handle to the first filespec match
   hFile = FindFirstFile(sPath, WFD)

  'if valid ...
   If hFile <> INVALID_HANDLE_VALUE Then

      Do

        'Even though this routine uses filespecs,
        '*.* is still valid and will cause the search
        'to return folders as well as files, so a
        'check against folders is still required.
         If Not (WFD.dwFileAttributes And FILE_ATTRIBUTE_DIRECTORY) _
                 = FILE_ATTRIBUTE_DIRECTORY Then

           'remove trailing nulls
            sTmp = TrimNull(WFD.cFileName)

         End If

      Loop While FindNextFile(hFile, WFD)

     'close the handle
      hFile = FindClose(hFile)

   End If

End Sub


Private Sub GetAllFilesSpecified(FP As FILE_PARAMS)

   Dim drvCount As Long
   Dim sBuffer As String
   Dim currDrive As String
   
   If FP.sFileRoot = "all fixed disks/partitions" Then
   
     'all drives
   
     'retrieve the available drives on the system
      sBuffer = Space$(64)
      drvCount = GetLogicalDriveStrings(Len(sBuffer), sBuffer)
   
     'drvCount returns the size of the drive string
      If drvCount Then
      
        'strip off trailing nulls
         sBuffer = Left$(sBuffer, drvCount)
              
        'search each drive for the file
         Do Until sBuffer = ""
   
           'strip off one drive item from sBuffer
            FP.sFileRoot = StripItem(sBuffer)
   
           'just search the local file system
            If GetDriveType(FP.sFileRoot) = DRIVE_FIXED Then
            
                             
               DoEvents
                              
               Call SearchForFilesArray(FP)
               
            End If
         
         Loop
      
      End If
      
   Else
   
       
       Call SearchForFilesArray(FP)
       
   End If

End Sub


Public Function TrimNull(startstr As String) As String

  'returns the string up to the first
  'null, if present, or the passed string
   Dim pos As Integer
   
   pos = InStr(startstr, Chr$(0))
   
   If pos Then
      TrimNull = Left$(startstr, pos - 1)
      Exit Function
   End If
  
   TrimNull = startstr
  
End Function


Private Sub SearchForFilesArray(FP As FILE_PARAMS)

  'local working variables
   Dim WFD As WIN32_FIND_DATA
   Dim hFile As Long
   Dim sPath As String
   Dim sRoot As String
   Dim sTmp As String
      
  'this routine is primarily interested in the
  'directories, so the file type must be *.*
   sRoot = QualifyPath(FP.sFileRoot)
   sPath = sRoot & "*.*"
   
  'obtain handle to the first match
   hFile = FindFirstFile(sPath, WFD)
   
  'if valid ...
   If hFile <> INVALID_HANDLE_VALUE Then
   
     'GetFileInformation function returns the number,
     'of files matching the filespec (FP.sFileNameExt)
     'in the passed folder.
      Call GetFileInformation(FP)

      Do
      
        'if the returned item is a folder...
         If (WFD.dwFileAttributes And FILE_ATTRIBUTE_DIRECTORY) Then
            
           'remove trailing nulls
            sTmp = TrimNull(WFD.cFileName)
            
           'and if the folder is not the default
           'self and parent folders...
            If sTmp <> "." And sTmp <> ".." Then
            
              'get the file
               FP.sFileRoot = sRoot & sTmp
               
               
              'This next If..Then just prevents adding extra
              'lines and unneeded paths to the array when a
              'file search is performed for a specific file type
               If FP.sFileNameExt = "*.*" Then
               
                 'Depending on the purpose, you may want to
                 'exclude the next 4 optional lines.
                 'The first two lines adds a blank entry
                 'to the array as a separator between new
                 'directories in the output file. The last
                 'two add the directory name alone, before
                 'listing the files underneath. These four
                 'lines can be optionally commented out).
                 'Obviously, these extra entries will skew
                 'the actual file counts.
                                 
               End If
               
              'call again
               Call SearchForFilesArray(FP)
            
            End If
               
            
         End If
         
     'continue looping until FindNextFile returns
     '0 (no more matches)
      Loop While FindNextFile(hFile, WFD)
      
     'close the find handle
      hFile = FindClose(hFile)
   
   End If
   
End Sub


Private Function QualifyPath(sPath As String) As String

  'assures that a passed path ends in a slash
   If Right$(sPath, 1) <> "\" Then
         QualifyPath = sPath & "\"
   Else: QualifyPath = sPath
   End If
      
End Function


Function StripItem(startStrg As String) As String

  'Take a string separated by Chr(0)'s, and split off 1 item, and
  'shorten the string so that the next item is ready for removal.
   Dim pos As Integer

   pos = InStr(startStrg, Chr$(0))

   If pos Then
      StripItem = Mid(startStrg, 1, pos - 1)
      startStrg = Mid(startStrg, pos + 1, Len(startStrg))
   End If

End Function



Private Sub GetSystemDrives(ctl As ComboBox)

   Dim drvCount As Long
   Dim sBuffer As String
   Dim currDrive As String
       
  'Retrieve the available drives on the system.
  'The string is padded with enough room to hold
  'the drives, nulls and final terminating null.
   sBuffer = Space$(105)
   drvCount = GetLogicalDriveStrings(Len(sBuffer), sBuffer)
   
  'drvCount returns the size of the drive string
   If drvCount Then
   
     'strip off trailing nulls
      sBuffer = Left$(sBuffer, drvCount)
           
     'search each drive for the file
      Do Until sBuffer = ""

        'strip off one drive item from sBuffer
         currDrive = StripItem(sBuffer)

        'just search the local file system
         If GetDriveType(currDrive) = DRIVE_FIXED Then
         
            ctl.AddItem Left$(currDrive, 2)
            
         End If
      
      Loop
      
   End If

End Sub

Public Sub RunShellExecute(sTopic As String, _
                            sFile As Variant, _
                            sParams As Variant, _
                            sDirectory As Variant, _
                            nShowCmd As Long)

   Dim hWndDesk As Long
   Dim success As Long
  
  'the desktop will be the
  'default for error messages
   hWndDesk = GetDesktopWindow()
  
  'execute the passed operation
   success = ShellExecute(hWndDesk, sTopic, sFile, sParams, sDirectory, nShowCmd)

  'have the "Open With.." dialog appear
  'when the ShellExecute API call fails
  If success < 32 Then
     Call Shell("rundll32.exe shell32.dll,OpenAs_RunDLL " & sFile, vbNormalFocus)
  End If
   
End Sub




Module6.bas
Attribute VB_Name = "Module6"
Option Explicit
Global FP As FILE_PARAMS
Global Itm As Node
Global favpath As String
Global sLastFolder As String
Global sRoot As String
Global bSubItem As Boolean
Global nCount As Long
Global bCancel As Boolean

Public Sub GetFavorites()
'///////////////////////////////////////////////////////////////
'FAVORITES LOAD
'    sLastFolder = ""
    'Load The Tree
    Form1.treeFavorites.Nodes.Clear
    Form1.treeFavorites.Refresh
    
    'retrieve the special folder path
    'to the internet favorites
    
    favpath = GetFolderPath(CSIDL_FAVORITES)
    
    'Initializes the Root Item in the TreeView
    Call LoadTreeView("Internet Favorites", True, True)
   
    If Len(favpath) > 0 Then
   
     'set up the search UDT
      With FP
         .sFileRoot = favpath
         .sFileNameExt = "*.url"
         .bRecurse = True
      End With
      
     'get the files
      Call SearchForFilesArray(FP)
      Form1.treeFavorites.Nodes("R").Expanded = True
    Else
         
       MsgBox " Could not locate favorites folder! " & _
           "This program requires Microsoft's Internet " & _
           "Explorer to be installed. Program will shutdown now!", _
           vbCritical + vbOKOnly, "FavMenu Error"
       End

    End If
'///////////////////////////////////////////////////////////////
End Sub

Public Function TrimNull(startstr As String) As String

  'returns the string up to the first
  'null, if present, or the passed string
   Dim pos As Integer
   
   pos = InStr(startstr, Chr$(0))
   
   If pos Then
      TrimNull = Left$(startstr, pos - 1)
      Exit Function
   End If
  
   TrimNull = startstr
  
End Function

Private Function GetFileInformation(FP As FILE_PARAMS) As Long

  'local working variables
   Dim WFD As WIN32_FIND_DATA
   Dim hFile As Long
   Dim pos As Long
   Dim sPath As String
   Dim sRoot As String
   Dim sTmp As String
   Dim sURL As String
   Dim sShortcut As String
   Dim itmX As ListItem
         
  'FP.sFileRoot (assigned to sRoot) contains
  'the path to search.
  '
  'FP.sFileNameExt (assigned to sPath) contains
  'the full path and filespec.
   sRoot = QualifyPath(FP.sFileRoot)
   sPath = sRoot & FP.sFileNameExt
   
  'obtain handle to the first filespec match
   hFile = FindFirstFile(sPath, WFD)
   
  'if valid ...
   If hFile <> INVALID_HANDLE_VALUE Then

      Do
      
        'remove trailing nulls
         sTmp = TrimNull(WFD.cFileName)
         
        'Even though this routine uses filespecs,
        '*.* is still valid and will cause the search
        'to return folders as well as files, so a
        'check against folders is still required.
         If Not (WFD.dwFileAttributes And FILE_ATTRIBUTE_DIRECTORY) _
            = FILE_ATTRIBUTE_DIRECTORY Then
           
           'determine the link name by removing
           'the .url extension
            pos = InStr(sTmp, ".url")
            
            If pos > 0 Then
            
                sShortcut = Left$(sTmp, pos - 1)
           
                'extract the URL
                sURL = ProfileGetItem("InternetShortcut", "URL", "", sRoot & sTmp)
                If sLastFolder = "" Then
                    'In The Root
                    Call LoadTreeView(sShortcut, False, False, "R", sURL)
                    'jrbCall LoadTreeView(sShortcut, False, False, "", sURL)
                Else
                    Call LoadTreeView(sShortcut, False, False, sLastFolder, sURL)
                End If
         
            End If
         
         End If
         
      Loop While FindNextFile(hFile, WFD)
      
     'close the handle
      hFile = FindClose(hFile)
   
   End If
   
  'clean up
   Set itmX = Nothing
   
End Function

Private Function QualifyPath(sPath As String) As String

  'assures that a passed path ends in a slash
   If Right$(sPath, 1) <> "\" Then
         QualifyPath = sPath & "\"
   Else: QualifyPath = sPath
   End If
      
End Function

Public Function GetFolderPath(CSIDL As Long) As String

   Dim sPath As String
   Dim sTmp As String
  
  'fill pidl with the specified folder item
   sPath = Space$(MAX_LENGTH)
   
   If SHGetFolderPath(Form1.hwnd, CSIDL, 0&, SHGFP_TYPE_CURRENT, sPath) = S_OK Then
       sTmp = Left$(sPath, InStr(sPath, Chr$(0)) - 1)
   End If
   
   GetFolderPath = sTmp
   
End Function

Public Function ProfileGetItem(lpSectionName As String, _
                               lpKeyName As String, _
                               defaultValue As String, _
                               inifile As String) As String

    'Retrieves a value from an ini file corresponding
    'to the section and key name passed.
          
     Dim success As Long
     Dim nSize As Long
     Dim ret As String
    
    'call the API with the parameters passed.
    'The return value is the length of the string
    'in ret, including the terminating null. If a
    'default value was passed, and the section or
    'key name are not in the file, that value is
    'returned. If no default value was passed (""),
    'then success will = 0 if not found.
    
    'Pad a string large enough to hold the data.
     ret = Space$(2048)
     nSize = Len(ret)
     success = GetPrivateProfileString(lpSectionName, lpKeyName, _
                                       defaultValue, ret, nSize, inifile)
     
     If success Then
        ProfileGetItem = Left$(ret, success)
     End If
     
End Function

Private Sub GetAllFilesSpecified(FP As FILE_PARAMS)

   Dim drvCount As Long
   Dim sBuffer As String
   Dim currDrive As String
   
   If FP.sFileRoot = "all fixed disks/partitions" Then
   
     'all drives
   
     'retrieve the available drives on the system
      sBuffer = Space$(64)
      drvCount = GetLogicalDriveStrings(Len(sBuffer), sBuffer)
   
     'drvCount returns the size of the drive string
      If drvCount Then
      
        'strip off trailing nulls
         sBuffer = Left$(sBuffer, drvCount)
              
        'search each drive for the file
         Do Until sBuffer = ""
   
           'strip off one drive item from sBuffer
            FP.sFileRoot = StripItem(sBuffer)
   
           'just search the local file system
            If GetDriveType(FP.sFileRoot) = DRIVE_FIXED Then
            
              'this may take a while, so update the
              'display when the search path changes
              'Text2.Text = "Working ... searching drive " & FP.sFileRoot
               
               DoEvents
               If bCancel Then Exit Do
               
               Call SearchForFilesArray(FP)
               
              'Update the display count
               'Text3.Text = Format$(nCount, sFileSoFar)
               'Text3.Refresh
               
            End If
         
         Loop
      
      End If
      
   Else
         
       Call SearchForFilesArray(FP)
       
   End If

End Sub

Private Sub SearchForFilesArray(FP As FILE_PARAMS)

  'local working variables
   Dim WFD As WIN32_FIND_DATA
   Dim hFile As Long
   Dim sPath As String
   Dim sRoot As String
   Dim sTmp As String
      
  'this routine is primarily interested in the
  'directories, so the file type must be *.*
   sRoot = QualifyPath(FP.sFileRoot)
   sPath = sRoot & "*.*"
   
  'obtain handle to the first match
   hFile = FindFirstFile(sPath, WFD)
   
  'if valid ...
   If hFile <> INVALID_HANDLE_VALUE Then
   
     'GetFileInformation function returns the number,
     'of files matching the filespec (FP.sFileNameExt)
     'in the passed folder.
      Call GetFileInformation(FP)

      Do
      
        'if the returned item is a folder...
         If (WFD.dwFileAttributes And FILE_ATTRIBUTE_DIRECTORY) Then
            
           'remove trailing nulls
            sTmp = TrimNull(WFD.cFileName)
            
           'and if the folder is not the default
           'self and parent folders...
            If sTmp <> "." And sTmp <> ".." Then
            
              
              'get the file
               FP.sFileRoot = sRoot & sTmp
              
              If InRoot(sTmp) Then
                Call LoadTreeView(sTmp, True, False)
                sLastFolder = sTmp
                  
              Else
                Call LoadTreeView(sTmp, True, False, sLastFolder)
                sLastFolder = sTmp
              End If
              
              'This next If..Then just prevents adding extra
              'lines and unneeded paths to the array when a
              'file search is performed for a specific file type
               If FP.sFileNameExt = "*.*" Then
               
                 'Depending on the purpose, you may want to
                 'exclude the next 4 optional lines.
                 'The first two lines adds a blank entry
                 'to the array as a separator between new
                 'directories in the output file. The last
                 'two add the directory name alone, before
                 'listing the files underneath. These four
                 'lines can be optionally commented out).
                 'Obviously, these extra entries will skew
                 'the actual file counts.
                  'nCount = nCount + 1
                  'sAllFiles(nCount) = ""
'                  nCount = nCount + 1
'
'                  sLastFolder = FP.sFileRoot
'                  sAllFiles(nCount) = FP.sFileRoot
                  
                  
               End If
               
              'call again
               Call SearchForFilesArray(FP)
            
            End If
               
            
         End If
         
     'continue looping until FindNextFile returns
     '0 (no more matches)
      Loop While FindNextFile(hFile, WFD)
      
     'close the find handle
      hFile = FindClose(hFile)
   
   End If
   
End Sub

Function StripItem(startStrg As String) As String

  'Take a string separated by Chr(0)'s, and split off 1 item, and
  'shorten the string so that the next item is ready for removal.
   Dim pos As Integer

   pos = InStr(startStrg, Chr$(0))

   If pos Then
      StripItem = Mid(startStrg, 1, pos - 1)
      startStrg = Mid(startStrg, pos + 1, Len(startStrg))
   End If

End Function

Private Sub GetSystemDrives(ctl As ComboBox)

   Dim drvCount As Long
   Dim sBuffer As String
   Dim currDrive As String
       
  'Retrieve the available drives on the system.
  'The string is padded with enough room to hold
  'the drives, nulls and final terminating null.
   sBuffer = Space$(105)
   drvCount = GetLogicalDriveStrings(Len(sBuffer), sBuffer)
   
  'drvCount returns the size of the drive string
   If drvCount Then
   
     'strip off trailing nulls
      sBuffer = Left$(sBuffer, drvCount)
           
     'search each drive for the file
      Do Until sBuffer = ""

        'strip off one drive item from sBuffer
         currDrive = StripItem(sBuffer)

        'just search the local file system
         If GetDriveType(currDrive) = DRIVE_FIXED Then
         
            ctl.AddItem Left$(currDrive, 2)
            
         End If
      
      Loop
      
   End If

End Sub
   
Private Function GetFolderName(ByVal sPath As String) As String
Dim length As Long
Dim xPos As Long
Dim sTemp As String

    GetFolderName = ""

    length = Len(sPath)
    xPos = length
    
    If Left(sPath, length) = "\" Then
        sPath = Left(sPath, (length - 1))
    End If
    
    Do Until xPos = 0
        xPos = xPos - 1
        
        If Mid$(sPath, xPos, 1) = "\" Then
            GetFolderName = Mid(sPath, (xPos - 1))
            Exit Do
        End If
        
    Loop
    
End Function

Private Sub LoadTreeView(ItemName As String, bFolder As Boolean, bRoot As Boolean, _
    Optional SubItem As String, Optional sURL As String)
        
    If bRoot Then
        Set Itm = Form1.treeFavorites.Nodes.Add(, , "R", ItemName, 4) '#image
        Itm.Tag = FP.sFileRoot
        Exit Sub
    End If
    On Error Resume Next
    If bFolder Then
        If Len(SubItem) > 0 Then
            Set Itm = Form1.treeFavorites.Nodes.Add(SubItem, tvwChild, ItemName, ItemName, 1, 2) '#image
            
        Else
            Set Itm = Form1.treeFavorites.Nodes.Add("R", tvwChild, ItemName, ItemName, 1, 2) '#image
            
        End If
        
        Itm.Tag = FP.sFileRoot
        
    Else
        If SubItem <> "VF Corporation" Then
            Set Itm = Form1.treeFavorites.Nodes.Add(SubItem, tvwChild, ItemName & "_URL", ItemName, 8) '#image
            Itm.Tag = sURL
        End If
        
    End If
    
End Sub

Private Function InRoot(ByVal sPath As String) As Boolean
Dim sTmp As String

    InRoot = False
    
    sTmp = favpath & "\" & sPath
    
    If Dir(sTmp, vbDirectory) <> "" Then
        InRoot = True
    End If
    
End Function



